### PR TITLE
feat: implement dump metrics

### PR DESCRIPTION
**Pull request by Igorgro**

### CODE REVIEW

File: substrate/substrate-dump/src/dumper.ts [hunk 1]
```diff
@@ -107,6 +114,7 @@ export class Dumper {
 
         let height = new Throttler(() => this.src().getFinalizedHeight(), 300_000)
         let chainHeight = await height.get()
+        this.prometheus().setLastBlock(this.options.lastBlock ? this.options.lastBlock : chainHeight);
```
Comment (eldargab) [inline substrate/substrate-dump/src/dumper.ts:117]: Metric should mean exactly one thing.

File: substrate/substrate-dump/package.json [hunk 2]
```diff
@@ -27,8 +27,10 @@
     "@subsquid/util-internal-counters": "^1.3.0",
     "@subsquid/util-internal-fs": "^0.1.0",
     "@subsquid/util-internal-hex": "^1.2.0",
+    "@subsquid/util-internal-prometheus-server": "^1.2.0",
     "@subsquid/util-internal-range": "^0.0.0",
-    "commander": "^11.0.0"
+    "commander": "^11.0.0",
+    "prom-client": "14.2.0"
```
Comment (eldargab) [inline substrate/substrate-dump/package.json:33]: Meaningless version pinning should not happen here. Use `^` specifier.

File: substrate/substrate-dump/src/main.ts [hunk 3]
```diff
@@ -22,6 +22,7 @@ runProgram(() => {
     program.option('--last-block <number>', 'Height of the last block to dump', nat)
     program.option('--with-trace [targets]', 'Fetch block trace')
     program.option('--chunk-size <MB>', 'Data chunk size in megabytes', positiveInt, 32)
+    program.option('--metrics-port <port>', 'Port to serve metrics on', positiveInt)
```
Comment (eldargab) [inline substrate/substrate-dump/src/main.ts:25]: `0` is a valid port number

Comment (eldargab) [reply substrate/substrate-dump/src/main.ts:25]: `program.option('--metrics <port>', 'Enable prometheus metrics', nat)`

File: substrate/substrate-dump/src/dumper.ts [hunk 4]
```diff
@@ -193,11 +201,14 @@ export class Dumper {
                 }
             }
         } else {
-            let archive = new ArchiveLayout(this.fs())
+            const archive = new ArchiveLayout(this.fs())
+            const prometheus = this.prometheus();
+            if (this.options.metricsPort) await prometheus.serve();
```
Comment (eldargab) [inline substrate/substrate-dump/src/dumper.ts:206]: Print `info` message about metrics

```ts
this.log().info(`prometheus metrics are available on port {port}`)
```

File: substrate/substrate-dump/src/dumper.ts [hunk 5]
```diff
@@ -193,11 +201,14 @@ export class Dumper {
...
             await archive.appendRawBlocks({
                 blocks: (nextBlock, prevHash) => this.saveMetadata(this.process(nextBlock, prevHash)),
                 range: this.range(),
-                chunkSize: this.options.chunkSize * 1024 * 1024
+                chunkSize: this.options.chunkSize * 1024 * 1024,
+                onSuccessWrite: this.options.metricsPort ? ((block) => { prometheus.setLastSavedBlock(block); }) : undefined
```
Comment (eldargab) [inline substrate/substrate-dump/src/dumper.ts:211]:
```ts
{
  onSuccessWrite: block => this.prometheus().setLastSavedBlock(block)
}
```

No need for availability check.

File: substrate/substrate-dump/src/prometheus.ts [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+    setLastBlock(block: number) {
+        this.lastBlockGauge.set(block);
+    }
+
+    setLastSavedBlock(block: number) {
+        this.lastSavedBlockGauge.set(block);
+    }
+
+    private getPort(): number | string {
```
Comment (eldargab) [inline substrate/substrate-dump/src/prometheus.ts:36]: Since `this.port` is always set, this method is NOOP.

**Review by eldargab — CHANGES_REQUESTED** [review]
Please, generate change files with `rush change -b origin/master`

File: substrate/substrate-dump/src/dumper.ts (hunk 1 above)
Comment (Igorgro) [reply substrate/substrate-dump/src/dumper.ts:117]: While I agree, I still not understand purpose of this metric when the `--last-block` specified

File: substrate/substrate-dump/src/dumper.ts (hunk 5 above)
Comment (Igorgro) [reply substrate/substrate-dump/src/dumper.ts:211]: Well, agree, but there is also no need to set metric if it is not used

File: util/util-internal-archive-layout/src/layout.ts [hunk 7]
```diff
@@ -190,6 +191,7 @@ export class ArchiveLayout {
                         assertNotNull(lastBlock)
                     ).transactDir('.', async fs => {
                         let content = await out.end()
+                        args.onSuccessWrite ? args.onSuccessWrite(lastBlock?.height || 0) : null;
```
[the remark this answers is not in the document]
Comment (eldargab) [reply util/util-internal-archive-layout/src/layout.ts]:
```ts
let blockRange = {
    from: assertNotNull(firstBlock),
    to: assertNotNull(lastBlock)
}

let chunk = getNextChunk(
    blockRange.from,
    blockRange.to
)

await chunk.transactDir('.', async fs => {})

args.onSuccessWrite?({
    blockRange,
    chunk: chunk.abs()
})
```

File: substrate/substrate-dump/src/prometheus.ts [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+export class PrometheusServer {
+    private registry = new Registry()
+    private port?: number | string
+    private lastBlockGauge: Gauge;
+    private lastSavedBlockGauge: Gauge;
+
+    constructor(port: number) {
+        this.port = port;
+        this.lastBlockGauge = new Gauge({
+            name: 'sqd_last_block_total',
```
[the remark this answers is not in the document]
Comment (eldargab) [reply substrate/substrate-dump/src/prometheus.ts:14]: SDK surely doesn't have `_total` suffixes, especially for non-aggregation metrics

File: substrate/substrate-dump/src/dumper.ts (hunk 1 above)
Comment (eldargab) [reply substrate/substrate-dump/src/dumper.ts:117]: RPC node can be behind `--last-block`

File: substrate/substrate-dump/src/dumper.ts [hunk 9]
```diff
@@ -203,12 +203,15 @@ export class Dumper {
...
+            if (this.options.metricsPort) {
+                await prometheus.serve();
+                this.log().info(`prometheus metrics are available on port {port}`)
+            }
             await archive.appendRawBlocks({
                 blocks: (nextBlock, prevHash) => this.saveMetadata(this.process(nextBlock, prevHash)),
                 range: this.range(),
                 chunkSize: this.options.chunkSize * 1024 * 1024,
-                onSuccessWrite: this.options.metricsPort ? ((block) => { prometheus.setLastSavedBlock(block); }) : undefined
+                onSuccessWrite: (block) => { prometheus.setLastSavedBlock(block); }
```
Comment (eldargab) [inline substrate/substrate-dump/src/dumper.ts:214]: Please, remove unneeded parentheses

```
block => prometheus.setLastSavedBlock(block)
```

**Comment by Igorgro** [comment]
> Please, generate change files with rush change -b origin/master

Done

**Comment by Igorgro** [comment]
Add rpc requests metrics

File: substrate/substrate-dump/src/dumper.ts [hunk 10]
```diff
@@ -92,10 +99,11 @@ export class Dumper {
                 : this.options.withTrace ? '' : undefined
         }
 
-        return this.src().getFinalizedBlocks([{
+        const blocks = this.src().getFinalizedBlocks([{
```
Comment (eldargab) [inline substrate/substrate-dump/src/dumper.ts:102]: Please, remove code style changes

File: substrate/substrate-dump/src/dumper.ts [hunk 11]
```diff
@@ -107,6 +115,7 @@ export class Dumper {
 
         let height = new Throttler(() => this.src().getFinalizedHeight(), 300_000)
         let chainHeight = await height.get()
+        this.prometheus().setChainHeight(chainHeight);
```
Comment (eldargab) [inline substrate/substrate-dump/src/dumper.ts:118]: No trailing `;`

File: substrate/substrate-dump/src/dumper.ts [hunk 12]
```diff
@@ -132,6 +141,9 @@ export class Dumper {
                     )
                 }
             }
+            const metrics = this.rpc().getMetrics();
+            this.prometheus().setSuccesfulRequestCount(metrics.requestsServed);
```
Comment (eldargab) [inline substrate/substrate-dump/src/dumper.ts:145]: This should be a pull based metric

File: substrate/substrate-dump/src/dumper.ts [hunk 13]
```diff
@@ -193,11 +205,17 @@ export class Dumper {
                 }
             }
         } else {
-            let archive = new ArchiveLayout(this.fs())
+            const archive = new ArchiveLayout(this.fs())
+            const prometheus = this.prometheus();
+            if (this.options.metricsPort) {
```
Comment (eldargab) [inline substrate/substrate-dump/src/dumper.ts:210]: `0` doesn't match this check, use `this.options.metricsPort == null`.

File: substrate/substrate-dump/src/dumper.ts [hunk 14]
```diff
@@ -193,11 +205,17 @@ export class Dumper {
...
+                await prometheus.serve();
+                this.log().info(`prometheus metrics are available on port ${this.options.metricsPort}`)
```
Comment (eldargab) [inline substrate/substrate-dump/src/dumper.ts:212]:
```ts
let promServer = await prometheus.serve()
this.log().info(`prometheus metrics are available on port ${promServer.port}`)
```

File: substrate/substrate-dump/src/main.ts [hunk 15]
```diff
@@ -22,6 +22,7 @@ runProgram(() => {
     program.option('--last-block <number>', 'Height of the last block to dump', nat)
     program.option('--with-trace [targets]', 'Fetch block trace')
     program.option('--chunk-size <MB>', 'Data chunk size in megabytes', positiveInt, 32)
+    program.option('--metrics-port <port>', 'Port to serve metrics on', nat)
```
Comment (eldargab) [inline substrate/substrate-dump/src/main.ts:25]:
```ts
program.option('--metrics <port>', 'Enable prometheus metrics server', nat)
```

File: util/util-internal-archive-layout/src/layout.ts [hunk 16]
```diff
@@ -185,12 +186,19 @@ export class ArchiveLayout {
...
+                    let blockRange = {
+                        from: assertNotNull(firstBlock),
+                        to: assertNotNull(lastBlock)
+                    }
+                    const chunk = getNextChunk(blockRange.from, blockRange.to)
+                    chunk.transactDir('.', async fs => {
                         let content = await out.end()
-                        return fs.write('blocks.jsonl.gz', content)
+                        const writeResult = fs.write('blocks.jsonl.gz', content)
+                        writeResult.then(() => args.onSuccessWrite ? args.onSuccessWrite({
```
Comment (eldargab) [inline util/util-internal-archive-layout/src/layout.ts:197]: Chunk is not yet written!

Please use optional chaining.

File: substrate/substrate-dump/src/dumper.ts [hunk 17]
```diff
@@ -82,6 +84,11 @@ export class Dumper {
         })
     }
 
+    @def
+    prometheus() {
+        return new PrometheusServer(this.options.metricsPort ?? 3000);
```
Comment (eldargab) [inline substrate/substrate-dump/src/dumper.ts:89]: `0` is a better default.

**Review by eldargab — CHANGES_REQUESTED** [review]
Please, avoid code style changes.

File: util/util-internal-archive-layout/src/layout.ts [hunk 18]
```diff
@@ -185,13 +186,19 @@ export class ArchiveLayout {
...
-                    await getNextChunk(
-                        assertNotNull(firstBlock),
-                        assertNotNull(lastBlock)
-                    ).transactDir('.', async fs => {
+                    let blockRange = {
+                        from: assertNotNull(firstBlock),
+                        to: assertNotNull(lastBlock)
+                    }
+                    const chunk = getNextChunk(blockRange.from, blockRange.to)
+                    chunk.transactDir('.', async fs => {
```
Comment (eldargab) [inline util/util-internal-archive-layout/src/layout.ts:194]: missing `await`

File: substrate/substrate-dump/src/prometheus.ts [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+        });
+
+        this.lastWrittenBlockGauge = new Gauge({
+            name: 'sqd_dump_last_written_block',
+            help: 'Last saved block',
+            registers: [this.registry]
+        });
+
+        this.rpcRequestsGauge = new Gauge({
+            name: 'sqd_rpc_requests_count',
```
[the remark this answers is not in the document]
Comment (Igorgro) [reply substrate/substrate-dump/src/prometheus.ts:28]: I see currently only one url. But ok, will add this

File: substrate/substrate-dump/src/prometheus.ts [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+
+        this.rpcRequestsGauge = new Gauge({
+            name: 'sqd_dump_rpc_requests_count',
+            help: 'Number of rpc requests of different kinds',
+            labelNames: ['kind', 'url'],
+            registers: [this.registry],
+            collect() {
+                const metrics = rpc.getMetrics();
+                this.set({
+                    kind: 'successful',
```
Comment (eldargab) [inline substrate/substrate-dump/src/prometheus.ts:35]: May be `success | failure` instead of `successful | failed`? Sounds better to me.

**Pull request merged**